src/main.cpp

```diff
@@ -2392,19 +2406,34 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
             BOOST_FOREACH(const uint256 &note_commitment, joinsplit.commitments) {
                 // Insert the note commitments into our temporary tree.
 
-                tree.append(note_commitment);
+                sprout_tree.append(note_commitment);
             }
         }
 
+        BOOST_FOREACH(const OutputDescription &outputDescription, tx.vShieldedOutput) {
+            sapling_tree.append(outputDescription.cm);
+        }
+
         vPos.push_back(std::make_pair(tx.GetHash(), pos));
         pos.nTxOffset += ::GetSerializeSize(tx, SER_DISK, CLIENT_VERSION);
     }
 
-    view.PushAnchor(tree);
+    view.PushSproutAnchor(sprout_tree);
+    view.PushSaplingAnchor(sapling_tree);
     if (!fJustCheck) {
-        pindex->hashAnchorEnd = tree.root();
+        pindex->hashFinalSproutRoot = sprout_tree.root();
+    }
+    blockundo.old_sprout_tree_root = old_sprout_tree_root;
+
+    // If Sapling is active, block.hashFinalSaplingRoot must be the
+    // same as the root of the Sapling tree
+    if (NetworkUpgradeActive(pindex->nHeight, chainparams.GetConsensus(), Consensus::UPGRADE_SAPLING)) {
+        if (block.hashFinalSaplingRoot != sapling_tree.root()) {
+            return state.DoS(100,
+                         error("ConnectBlock(): block's hashFinalSaplingRoot is incorrect"),
+                               REJECT_INVALID, "bad-sapling-root-in-block");
```

Comment:
Sanity check: Does this error happen at the correct place, relative to the side-effects of `ConnectBlock`? Will all side-effects prior to this call be reversed?

Reply:
Yup. After this check, we even still check `if (block.vtx[0].GetValueOut() > blockReward)`.